src/filesystem.ts

```diff
@@ -0,0 +1,70 @@
+import localforage from 'localforage'
+
+import FileSystem from './fs'
+import * as cidLog from './common/cid-log'
+import * as dataRoot from './data-root'
+import { authenticatedUsername } from './common'
+
+
+/**
+ * Load a user's file system.
+ *
+ * @param username Optional, username of the user to load the file system from.
+ *                 Will try to load the file system of the authenticated user
+ *                 by default. Throws an error if there's no authenticated user.
+ */
+export async function loadFileSystem(username?: string): Promise<FileSystem> {
+  let cid, fs
+
+  // Look for username
+  username = username || (await authenticatedUsername() || undefined)
+  if (!username) throw new Error("User hasn't authenticated yet")
+
+  // Determine the correct CID of the file system to load
+  const dataCid = await dataRoot.lookup(username)
+  const [ logIdx, logLength ] = dataCid ? await cidLog.index(dataCid) : [ -1, 0 ]
+
+  if (!dataCid) {
+    // No DNS CID yet
+    cid = await cidLog.newest()
+
+  } else if (logIdx === 0) {
+    // DNS is up to date
+    if (logLength > 1) await cidLog.override(dataCid)
+    cid = dataCid
+
+  } else if (logIdx > 0) {
+    // DNS is outdated
+    cidLog.removeOlderCids(logIdx)
+    cid = await cidLog.newest()
+
+  } else {
+    // DNS is newer
+    cid = dataCid
+
+  }
```

Comment:
Implemented that CID log @expede 
Here you can see all the conditions it checks for.

Reply:
This feels like a lot of steps. Why not something like:

```js
if (cidLog.includes(dnsRoot)) {
  // Local ahead of remote
  cidLog.push(dnsRoot)
} else {
  // Remote ahead of local
  cidLog = [dnsRoot]
  localRoot = dnsRoot
}
```

Reply:
Specifically, we shouldn’t care about ordering, so `=== 0` vs `> 0` shouldn’t matter. Because DNS is eventually consistent, you may get out of order reads from it. Set membership is probably safer/easier, unless I’m missing something

Reply:
Speaking of sets, may as well use a `Set` rather than an array, since we don’t care about order

Reply:
I don't think a `Set` works here. What if your local file system is multiple CIDs ahead and you refresh the page, how do you know which is the last one? You are correct that I could eliminate one branch here and just use `log.includes()`

`setDNS(dnsRoot)` would conflict with other code yeah. Because a DNS update will always be done for every file-system change, it might just not be persisted yet.

> Because DNS is eventually consistent, you may get out of order reads from it.

Which is something that we can take advantage of when have versioning and/or crdt stuff? Not right now I assume?

Reply:
> What if your local file system is multiple CIDs ahead and you refresh the page, how do you know which is the last one

It shouldn’t matter because all the cache says is “I’ve seen this one before. The one that I have in localStorage is thus up to date.” The cache only has the single responsibility: to check if this is something new. The filesystem can’t be rewound. With an array, we have an implicit assumption that the first one is the current head of the history. Until we support versioning, we don’t care about the order earlier than this.

I think maybe the main difference between our approaches is I’m assuming that the current filesystem root is kept in a different variable than this cache. You _can_ overload them, but then you need to check for the there states.

---

> Which is something that we can take advantage of when we have versioning and/or crdt stuff? Not right now I assume?

It’s never going to be an asset, only a liability. What CRDTs, Merkle history, and whatnot give us is a way of not being dependent on time or order — our data dependencies are only the things that matter to the operation of the program. DNS will sometimes return old results, which is weird for humans to think about because we want a nice linear ordering with a clear “now”, but distributed systems (DNS, CRDTs, GenServer retry) are a bit more “quantum”.

Reply:
> I think maybe the main difference between our approaches is I’m assuming that the current filesystem root is kept in a different variable than this cache.

Oh ok, yes that's indeed the difference. Makes sense now 👍 

> I love talking about this stuff, but writing a small textbook here isn’t the more effective way to communicate the topic in full. Always happy to hop on a call to explain with more hand gestures :P

Yes please 😄 Lets talk about it later during our call. I understand the situation and liabilities, but don't know about the techniques to use to fix this.